File: src/pocketcloud/plugin/CloudPluginManager.php
<?php

namespace pocketcloud\plugin;

use pocketcloud\event\EventManager;
use pocketcloud\event\impl\plugin\PluginDisableEvent;
use pocketcloud\event\impl\plugin\PluginEnableEvent;
use pocketcloud\event\impl\plugin\PluginLoadEvent;
use pocketcloud\language\Language;
use pocketcloud\plugin\loader\FolderCloudPluginLoader;
use pocketcloud\plugin\loader\PharCloudPluginLoader;
use pocketcloud\plugin\loader\CloudPluginLoader;
use pocketcloud\util\CloudLogger;
use pocketcloud\util\Reloadable;
use pocketcloud\util\SingletonTrait;
use pocketcloud\util\Tickable;

class CloudPluginManager implements Tickable, Reloadable {
    use SingletonTrait;

    /** @var array<CloudPlugin> */
    private array $plugins = [];
    /** @var array<CloudPlugin> */
    private array $enabledPlugins = [];
    /** @var array<CloudPluginLoader> */
    private array $loaders = [];

    public function __construct() {
        self::setInstance($this);
        $this->registerLoader(new PharCloudPluginLoader());
        $this->registerLoader(new FolderCloudPluginLoader());
    }

    public function registerLoader(CloudPluginLoader $loader): void {
        $this->loaders[] = $loader;
    }

    public function loadPlugins(): void {
        CloudLogger::get()->info(Language::current()->translate("plugins.loading"));
        foreach (array_diff(scandir(CLOUD_PLUGINS_PATH), [".", ".."]) as $file) {
            $path = CLOUD_PLUGINS_PATH . $file;
            $this->loadPlugin($path);
        }

        if (count($this->plugins) == 0) {
            CloudLogger::get()->info(Language::current()->translate("plugins.loaded.none"));
        } else {
            CloudLogger::get()->info(Language::current()->translate("plugins.loaded", count($this->plugins)));
        }
    }

    public function loadPlugin(string $path): void {
        CloudLogger::get()->info(Language::current()->translate("plugin.load", basename($path)));
        foreach ($this->loaders as $loader) {
            try {
                if ($loader->canLoad($path)) {
                    $plugin = $loader->loadPlugin($path);
                    if (!$plugin instanceof CloudPlugin) {
                        CloudLogger::get()->error(Language::current()->translate("plugin.loading.failed", basename($path), $plugin));
                        return;
                    }

                    if (isset($this->plugins[$plugin->getDescription()->getName()])) {
                        CloudLogger::get()->error(Language::current()->translate("plugin.loading.failed", $plugin->getDescription()->getName(), "Plugin already exists"));
                        return;
                    }

                    (new PluginLoadEvent($plugin))->call();
                    $this->plugins[$plugin->getDescription()->getName()] = $plugin;
                    $plugin->onLoad();
                }
            } catch (\Throwable $exception) {
                CloudLogger::get()->error(Language::current()->translate("plugin.loading.failed", basename($path), $exception->getMessage()));
                CloudLogger::get()->exception($exception);
            }
        }
    }

    public function enablePlugins(): void {
        CloudLogger::get()->info(Language::current()->translate("plugins.enabling"));
        foreach ($this->plugins as $plugin) {
            $this->enablePlugin($plugin);
        }

        if (count($this->enabledPlugins) == 0) {
            CloudLogger::get()->info(Language::current()->translate("plugins.enabled.none"));
        } else {
            CloudLogger::get()->info(Language::current()->translate("plugins.enabled", count($this->enabledPlugins)));
        }
    }

    public function enablePlugin(CloudPlugin $plugin): void {
        CloudLogger::get()->info(Language::current()->translate("plugin.enabling", $plugin->getDescription()->getName(), $plugin->getDescription()->getFullName()));
        $plugin->setEnabled(true);
        (new PluginEnableEvent($plugin))->call();
        try {
            $plugin->onEnable();
        } catch (\Throwable $throwable) {
            CloudLogger::get()->exception($throwable);
            $this->disablePlugin($plugin);
        }

        if ($plugin->isEnabled()) {
            $this->enabledPlugins[$plugin->getDescription()->getName()] = $plugin;
        }
    }

    public function disablePlugins(): void {
        foreach ($this->enabledPlugins as $plugin) {
            $this->disablePlugin($plugin);
        }

        CloudLogger::get()->info(Language::current()->translate("plugins.disabled"));
    }

    public function disablePlugin(CloudPlugin $plugin): void {
        CloudLogger::get()->info(Language::current()->translate("plugin.disabling", $plugin->getDescription()->getName(), $plugin->getDescription()->getFullName()));
        (new PluginDisableEvent($plugin))->call();
        $plugin->setEnabled(false);
        $plugin->onDisable();

        $plugin->getScheduler()->cancelAll();
        EventManager::getInstance()->removeHandlers($plugin);
        if (isset($this->enabledPlugins[$plugin->getDescription()->getName()])) unset($this->enabledPlugins[$plugin->getDescription()->getName()]);
    }

    public function clear(): void {
        $this->plugins = [];
        $this->enabledPlugins = [];
    }

    public function reload(): bool {
        foreach ($this->plugins as $plugin) {
            if ($plugin->isEnabled()) $this->disablePlugin($plugin);
            unset($this->plugins[$plugin->getDescription()->getName()]);
        }

        $this->clear();

        foreach (array_diff(scandir(CLOUD_PLUGINS_PATH), [".", ".."]) as $file) {
            $path = CLOUD_PLUGINS_PATH . $file;
            $this->loadPlugin($path);
        }
        return true;
    }

    public function tick(int $currentTick): void {
        foreach ($this->enabledPlugins as $enabledPlugin) {
            if ($enabledPlugin->isEnabled()) {
                $enabledPlugin->getScheduler()->tick($currentTick);
            }
        }
    }

    public function getPluginByName(string $name): ?CloudPlugin {
        return $this->plugins[$name] ?? null;
    }

    public function getLoaders(): array {
        return $this->loaders;
    }

    public function getEnabledPlugins(): array {
        return $this->enabledPlugins;
    }

    public function getPlugins(): array {
        return $this->plugins;
    }

    public static function getInstance(): ?self {
        return self::$instance;
    }
}

File: src/main/java/de/pocketcloud/cloud/plugin/CloudPluginManager.java
package de.pocketcloud.cloud.plugin;

import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.event.impl.plugin.PluginDisableEvent;
import de.pocketcloud.cloud.event.impl.plugin.PluginEnableEvent;
import de.pocketcloud.cloud.event.impl.plugin.PluginLoadEvent;
import de.pocketcloud.cloud.load.Loadable;
import de.pocketcloud.cloud.plugin.exception.PluginLoadFailedException;
import de.pocketcloud.cloud.plugin.loader.JarCloudPluginLoader;
import de.pocketcloud.cloud.tick.Tickable;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.io.IOException;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Getter
public final class CloudPluginManager implements Tickable, Loadable {

    @Getter
    @Accessors(fluent = true)
    private static CloudPluginManager instance = null;

    private final Map<String, CloudPlugin> plugins = new HashMap<>();
    private final JarCloudPluginLoader pluginLoader = new JarCloudPluginLoader();
    private final Path pluginsFolder = Paths.get("storage/plugins");

    public CloudPluginManager() {
        instance = this;
    }

    public void load() {
        CloudLogger.get().info("Loading plugins...");
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(pluginsFolder)) {
            for (Path path : stream) {
                if (Files.isRegularFile(path)) {
                    if (path.getFileName().toString().endsWith(".jar")) {
                        loadPlugin(path);
                    }
                }
            }
        } catch (IOException e) {
            CloudLogger.get().exception("Unable to load plugins", e);
        } finally {
            enableAll();
        }
    }

    @Override
    public void unload() {
        disableAll();
        this.plugins.clear();
    }

    public void loadPlugin(Path jarFile) {
        if (pluginLoader.canLoad(jarFile)) {
            try {
                CloudLogger.get().info("Loading plugin §b{}§r...", jarFile.getFileName().toString());
                CloudPlugin plugin = pluginLoader.load(jarFile);
                if (plugins.containsKey(plugin.getDescription().name())) throw new PluginLoadFailedException("Plugin with the same name already loaded");
                plugins.put(plugin.getDescription().name(), plugin);
                new PluginLoadEvent(plugin).call();
                plugin.onLoad();
            } catch (Exception e) {
                CloudLogger.get().exception("Unable to load plugin " + jarFile.getFileName(), e);
            }
        }
    }

    public void enableAll() {
        plugins.values().stream().filter(CloudPlugin::isDisabled).forEach(this::enable);
    }

    public void enable(CloudPlugin plugin) {
        if (plugin.isEnabled()) return;
        CloudLogger.get().info("§aEnabling §rplugin §b{}§r...",  plugin.getDescription().name());
        plugin.setState(CloudPluginState.ENABLED);
        try {
            new PluginEnableEvent(plugin).call();
            plugin.onEnable();
        } catch (Exception e) {
            plugin.getLogger().exception("Unable to enable plugin", e);
            disable(plugin);
        }
    }

    public void disableAll() {
        CloudLogger.get().info("§cDisabling §rall plugins...");
        plugins.values().stream().filter(CloudPlugin::isEnabled).forEach(this::disable);
    }

    public void disable(CloudPlugin plugin) {
        if (plugin.isDisabled()) return;
        CloudLogger.get().info("§cDisabling §rplugin §b{}§r...",  plugin.getDescription().name());
        plugins.remove(plugin.getDescription().name());
        plugin.setState(CloudPluginState.DISABLED);
        try {
            new PluginDisableEvent(plugin).call();
            plugin.onDisable();
        } catch (Exception e) {
            plugin.getLogger().exception("Unable to disable plugin", e);
        } finally {
            try {
                plugin.getLoader().close();
            } catch (Exception _) {}
        }
    }

    @Override
    public void tick(long currentTick) {
        for (CloudPlugin plugin : plugins.values()) {
            if (!plugin.isEnabled()) continue;
            plugin.getScheduler().tick(currentTick);
        }
    }

    public Optional<CloudPlugin> get(String name) {
        return Optional.ofNullable(plugins.getOrDefault(name, null));
    }

    public List<CloudPlugin> getEnabledPlugins() {
        return plugins.values().stream().filter(CloudPlugin::isEnabled).collect(Collectors.toList());
    }

    public List<CloudPlugin> getDisabledPlugins() {
        return plugins.values().stream().filter(CloudPlugin::isDisabled).collect(Collectors.toList());
    }
}